APMrover2/AP_MotorsUGV.cpp

```diff
@@ -0,0 +1,256 @@
+
+    // output to throttle channels
+    if (armed) {
+        // handle armed case
+        SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, throttle);
+    } else {
+        // handle disarmed case
+        if (_disarm_disable_pwm) {
+            SRV_Channels::set_output_limit(SRV_Channel::k_throttle, SRV_Channel::SRV_CHANNEL_LIMIT_ZERO_PWM);
+        } else {
+            SRV_Channels::set_output_limit(SRV_Channel::k_throttle, SRV_Channel::SRV_CHANNEL_LIMIT_TRIM);
+        }
+    }
+}
+
+// output to skid steering channels
+void AP_MotorsUGV::output_skid_steering(bool armed, float steering, float throttle)
+{
+    // handle simpler disarmed case
+    if (!armed) {
+        if (_disarm_disable_pwm) {
+            SRV_Channels::set_output_limit(SRV_Channel::k_throttleLeft, SRV_Channel::SRV_CHANNEL_LIMIT_ZERO_PWM);
+            SRV_Channels::set_output_limit(SRV_Channel::k_throttleRight, SRV_Channel::SRV_CHANNEL_LIMIT_ZERO_PWM);
+        } else {
+            SRV_Channels::set_output_limit(SRV_Channel::k_throttleLeft, SRV_Channel::SRV_CHANNEL_LIMIT_TRIM);
+            SRV_Channels::set_output_limit(SRV_Channel::k_throttleRight, SRV_Channel::SRV_CHANNEL_LIMIT_TRIM);
+        }
+        return;
+    }
+
+    // skid steering mixer
+    float steering_scaled = steering / 4500.0f; // steering scaled -1 to +1
+    float throttle_scaled = throttle / 100.0f;  // throttle scaled -1 to +1
+
+    // apply constraints
+    steering_scaled = constrain_float(steering_scaled, -1.0f, 1.0f);
+    throttle_scaled = constrain_float(throttle_scaled, -1.0f, 1.0f);
+
+    // check for saturation and scale back throttle and steering proportionally
+    const float saturation_value = fabsf(steering_scaled) + fabsf(throttle_scaled);
+    if (saturation_value > 1.0f) {
+        steering_scaled = steering_scaled / saturation_value;
+        throttle_scaled = throttle_scaled / saturation_value;
+    }
+
+    // add in throttle
+    float motor_left = throttle_scaled;
+    float motor_right = throttle_scaled;
+
+    // deal with case of turning on the spot
+    if (is_zero(throttle_scaled)) {
+        // full possible range is not used to keep response equivalent to non-zero throttle case
+        motor_left += steering_scaled * 0.5f;
+        motor_right -= steering_scaled * 0.5f;
+    } else {
+        // add in steering
+        const float dir = is_positive(throttle_scaled) ? 1.0f : -1.0f;
+        if (is_negative(steering_scaled)) {
+            // moving left all steering to right wheel
+            motor_right -= dir * steering_scaled;
+        } else {
+            // turning right, all steering to left wheel
+            motor_left += dir * steering_scaled;
+        }
+    }
+
+    SRV_Channels::set_output_scaled(SRV_Channel::k_throttleLeft,  1000.0f * motor_left);
+    SRV_Channels::set_output_scaled(SRV_Channel::k_throttleRight, 1000.0f * motor_right);
+}
+
+// slew limit throttle for one iteration
+void AP_MotorsUGV::slew_limit_throttle(float dt)
```

Comment:
We are limiting the throttle but not the steering. In case of pivot turn, we will have big response.